Add introductory text for models in documentation

jessica-mitchell opened the pull request:
This PR adds introductory text to explain a bit about models in NEST.

@clinssen @pnbabu I added a section on NESTML that is copied from your docs, please comment if you have some other text you would like better.

There is a table of terms used in model names along with commonly used terms in literature. It's incomplete and in need of a model expert to help fill it out.

https://nest-test.readthedocs.io/en/add-intro-models/models/models-main.html

clinssen commented on doc/htmldoc/models/models-main.rst:
```diff
@@ -1,12 +1,152 @@
+
 NEST provides a ton of models! Textbook standards like integrate-and-fire and Hodgkin-Huxley type models are available
 alongside high quality implementations of models published by the neuroscience community.
+The models are organized and autogenerated by keywords (e.g., adaptive threshold, conductance-based etc) in our directory. Models that contan a specific keyword will be listed under that word.
```
can keyword examples straightaway be links?

jessica-mitchell replied:
yes

clinssen commented on doc/htmldoc/models/models-main.rst:
```diff
@@ -1,12 +1,152 @@
+
+* Discover :doc:`all the models in our directory <index>`.
+
+NESTML
```
Maybe this paragraph as an admonition (``.. hint``)?

jessica-mitchell replied:
done

clinssen commented on doc/htmldoc/models/models-main.rst:
```diff
@@ -1,12 +1,152 @@
+NESTML
+------
+
+* Check out :doc:`NESTML <nestml:index>`,  a domain-specific language for neuron and synapse models. These dynamical models can
```
```suggestion
* Check out :doc:`NESTML <nestml:index>`, a domain-specific language for neuron and synapse models. NESTML enables fast prototyping of new models using an easy to understand, yet powerful syntax, and combines this with high simulation performance through the automated generation of C++ code, suitable for use in NEST Simulator.
```

clinssen reviewed (commented):
Just some initial ideas!

pnbabu commented on doc/htmldoc/models/models-main.rst:
```diff
@@ -1,12 +1,152 @@
+* fast prototyping
+* an easy to understand, yet powerful syntax;
+* a flexible processing toolchain, written in Python;
+* good simulation performance by means of code generation (C++ for NEST Simulator).
```
Can we also give some instructions on how to install NESTML here or on the install page?

jessica-mitchell replied:
done

pnbabu commented on doc/htmldoc/models/models-main.rst:
```diff
@@ -1,12 +1,152 @@
+What we mean by `models`
+------------------------
+
+Models in the context of NEST are C++ implmentations of mathematical equations that describe the characteristics and behavior of
```
```suggestion
Models in the context of NEST are C++ implementations of mathematical equations that describe the characteristics and behavior of
```

pnbabu commented on doc/htmldoc/models/models-main.rst:
```diff
@@ -1,12 +1,152 @@
+------------------------
+
+Models in the context of NEST are C++ implmentations of mathematical equations that describe the characteristics and behavior of
+different types of neurons, syanpses, (and devices?), based on the relevant peer-reviewed publications for the model.
```
```suggestion
different types of neurons, synapses, (and devices?), based on the relevant peer-reviewed publications for the model.
```

pnbabu commented on doc/htmldoc/models/models-main.rst:
```diff
@@ -1,12 +1,152 @@
+different types of neurons, syanpses, (and devices?), based on the relevant peer-reviewed publications for the model.
+
+We also use the term model in relation to network models (e.g., :ref:`microcircuit <toc_microcircuit>` and `multi-area model <https://inm-6.github.io/multi-area-model/>`_). These network models
+can be considerd a level of complexity higher than the neuron or synapse model. We do not focus on network models here.
```
```suggestion
can be considered a level of complexity higher than the neuron or synapse model. We do not focus on network models here.
```

pnbabu commented on doc/htmldoc/models/models-main.rst:
```diff
@@ -1,12 +1,152 @@
+
 NEST provides a ton of models! Textbook standards like integrate-and-fire and Hodgkin-Huxley type models are available
 alongside high quality implementations of models published by the neuroscience community.
+The models are organized and autogenerated by keywords (e.g., adaptive threshold, conductance-based etc) in our directory. Models that contan a specific keyword will be listed under that word.
```
```suggestion
The models are organized and autogenerated by keywords (e.g., adaptive threshold, conductance-based etc) in our directory. Models that contain a specific keyword will be listed under that word.
```

pnbabu commented on doc/htmldoc/models/models-main.rst:
```diff
@@ -1,12 +1,152 @@
+Synapse models
+~~~~~~~~~~~~~~
+
+Synapse models  include the word synapse as last word in model name.
```
```suggestion
Synapse models include the word synapse as last word in the model name.
```

jessica-mitchell commented:
@pnbabu @clinssen Thanks for your comments, I applied your suggestions, and I created a page on the installation for installing NEST with NESTML - it doesnt got into much detail. I think most of the information should be kept in your docs - the only thing that might be necessary imo, is if there is any additional setup to get NESTML to work with NEST aside from having them both installed. Also note that in #2289 I added a link to the global toc for NESTML.

terhorstd commented:
@jessica-mitchell, the last merges caused a tiny conflict, could you resolve those? This seems otherwise almost ready to go… don't hesitate to re-request review when appropriate :)

jessica-mitchell commented:
@terhorstd I'm still looking for someone to review the table I created that compares NEST terms with commonly used ones in research (like CUBA and psc for current-based)

clinssen commented on doc/htmldoc/models/models-main.rst:
```diff
@@ -1,12 +1,141 @@
+The following table provides the NEST term used for a specific descriptor of a model along with commonly used alternative terms you may find in the literature.
+Additional terms may be used to help differentiate models not included in the list below.
+See the the :doc:`model directory <index>` to access the full description of each model.
+
```
link to glossary instead/additionally?

jessica-mitchell replied:
yes, I think instead of table here it would better to put terms in a subsection of glossary and link to that.

jessica-mitchell replied:
@clinssen I reorganized glossary into subsections and linked to the a section on *terms for models*

jessica-mitchell commented on doc/htmldoc/models/models-main.rst:
```diff
@@ -1,12 +1,141 @@
+~~~~~~~
+
+A device name should represent its physical counter point - like a multimeter is ``multimeter``.  In general, the term ``recorder`` is used for devices
+that store or generate output. The term  ``generator`` is used for devices that provide input into the simulation.
```
Couldnt the output be various forms of data? would providing a couple of examples work?

clinssen replied:
sure, perhaps something like "output such as spike times or synaptic strengths over time"?

jessica-mitchell replied:
done

jessica-mitchell commented on doc/htmldoc/models/models-main.rst:
```diff
@@ -1,12 +1,141 @@
+Devices
+~~~~~~~
+
+A device name should represent its physical counter point - like a multimeter is ``multimeter``.  In general, the term ``recorder`` is used for devices
```
d'oh, yes I'll fix that

jessica-mitchell commented on doc/htmldoc/installation/nestml.rst:
```diff
@@ -0,0 +1,27 @@
+.. _nestml:
+
+Install NESTML
```
yah I dont quite like this, but there was discussion about making nestml more visible in the installation section of nest so I was trying to utilize that. In this case, maybe a direct link to nestml is better

clinssen commented on doc/htmldoc/installation/index.rst:
```diff
@@ -29,8 +29,16 @@ Install NEST
+Install NESTML + NEST
+---------------------
+
+|nestml| I'm a user who wants to :ref:`create or customize models <nestml>`
```
```suggestion
|nestml| I'm a user who wants to :ref:`create or customize neuron and synapse models <nestml>`
```

clinssen replied:
The "|nestml|" part does not seem to render well on RtD. Suggest to link directly to https://nestml.readthedocs.org/

clinssen replied:
Suggest to remove "custom models" from the left-hand side menu as it links outside of the NEST RtD documentation which can be confusing.

jessica-mitchell replied:
Ok I will directy link to nestml.
On an aside, do you have information for installing nestml with nest? we have docker that contains nestml and nest, but if you install the two separately how does a user ensure they work together? This is something not for this PR, but we should make sure it's clear to users.


I think we should  keep custom models: this was something discussed with Jochen  as well, and wanting to highlight nestml more. I looked to find a way to include an icon for external links using the new theme but couldn't find anything yet.

jessica-mitchell replied:
see here for current install page of nest: https://nest-test.readthedocs.io/en/add-intro-models/installation/index.html

pnbabu commented on doc/htmldoc/models/models-main.rst:
```diff
@@ -1,12 +1,93 @@
+
+Synapse models include the word synapse as last word in the model name.
+
+Synapse models may begin with author name (e.g., ``clopath_synapsae``) or process (e.g., ``stdp_synase``).
```
```suggestion
Synapse models may begin with author name (e.g., ``clopath_synapse``) or process (e.g., ``stdp_synapse``).
```

pnbabu commented on doc/htmldoc/models/models-main.rst:
```diff
@@ -1,12 +1,93 @@
+.. seealso::
 
-Need to create your own?
+  See our glossary section on :ref:`common abbreviations used for model terms <model_terms>`. It includes alternative terms commonly used in the literatue.
```
```suggestion
  See our glossary section on :ref:`common abbreviations used for model terms <model_terms>`. It includes alternative terms commonly used in the literature.
```

pnbabu reviewed (commented):
Some additional comments
- In the `Lecturer installation instructions` page, under the `Docker image` section, could you also mention that NESTML is included within the NEST docker image?
- Can the `nest::docs` logo in the header be made clickable so that the user can navigate to the home page by clicking that?
- The `Contents` on the right-hand side for all pages contain a link to `Show Source` . Is this kept deliberately?

jessica-mitchell commented on doc/extractor_userdocs.py:
```diff
@@ -131,7 +131,7 @@ def UserDocExtractor(
                 doc = rewrite_see_also(doc, filename, tags)
             except ValueError as e:
-                log.warning("Failed to rebuild 'See also' section: %s", e)
+                log.info("Failed to rebuild 'See also' section: %s", e)
```
I did this because I wanted to reduce the number of output messages that get generated during build, and I dont think these need to be warnings

jessica-mitchell replied:
The problem is that this warning is spit out I think for nearly all models, and it's an issue more with the script I think than the model generation itself. (See also sections seem to be created correctly at least for the most part) We do need to look into this and fix the underlying issues in another PR. But for now I wanted to make the build a little bit cleaner looking.

akorgor replied:
Ok

jessica-mitchell commented:
> Some additional comments
> 
>     * In the `Lecturer installation instructions` page, under the `Docker image` section, could you also mention that NESTML is included within the NEST docker image?

Done
> 
>     * Can the `nest::docs` logo in the header be made clickable so that the user can navigate to the home page by clicking that?
Yes, but that's for another PR. 
>     * The `Contents` on the right-hand side for all pages contain a link to `Show Source` . Is this kept deliberately?
That is the default for the theme and I dont see anything wrong with it. If you think this needs to change then please create an issue for it. I don't want to add a bunch more irrelevant things to this PR or it will never be merged. The focus is on the model intro page.

clinssen reviewed (approved):
Thanks!

akorgor commented on doc/htmldoc/ref_material/glossary.rst:
```diff
@@ -134,6 +88,90 @@ Glossary
+ :sorted:
+
+ iaf
+   Integrate and fire. Also known in other sources as `IF`.
```
```suggestion
   Integrate-and-fire. Also known in other sources as `IF`.
```

akorgor commented on doc/htmldoc/ref_material/glossary.rst:
```diff
@@ -134,6 +88,90 @@ Glossary
+   Generalized integrate and fire. From Gersnter lab.
+
+ glif
+   Generalized leaky integrate and fire. From Allen institute.
```
```suggestion
   Generalized leaky integrate-and-fire. From the Allen institute.
```

akorgor commented on doc/htmldoc/ref_material/glossary.rst:
```diff
@@ -134,6 +88,90 @@ Glossary
+   Excitatory.
+
+ stdp
+   Spike-timing dependent plasticity synapse.
```
```suggestion
   Spike-timing dependent plasticity.
```

akorgor replied:
To make it consistent with the short-term plasticity entry.

akorgor commented on doc/htmldoc/ref_material/glossary.rst:
```diff
@@ -134,6 +88,90 @@ Glossary
+   Spike-timing dependent plasticity synapse.
+
+ st
+   Short term plasticity.
```
```suggestion
   Short-term plasticity.
```

akorgor commented on doc/htmldoc/ref_material/glossary.rst:
```diff
@@ -134,6 +88,90 @@ Glossary
+   Spike-frequency adaptation.
+
+ cm
+   compartmental model.
```
```suggestion
   Compartmental model.
```

akorgor replied:
For consistency.

akorgor commented on doc/htmldoc/ref_material/glossary.rst:
```diff
@@ -256,12 +288,6 @@ Glossary
- PSC
-   Post-synatpic current.
-
  absolute refractory
```
```suggestion
 absolute refractory period
```

akorgor commented on doc/htmldoc/ref_material/glossary.rst:
```diff
@@ -256,12 +288,6 @@ Glossary
-   Post-synatpic current.
-
  absolute refractory
    An interval after a neuron fires a spike to prevent it from firing a spike again.
```
```suggestion
   Interval directly following a spike emission in which the sender neuron cannot fire again.
```

akorgor replied:
The reason for the refractory period is of biological nature (depletion of neurotransmitters etc.). As the description was formulated it sounds as if there was a technical reason we would need to prevent follow-up spikes.

akorgor commented on doc/htmldoc/models/models-main.rst:
```diff
@@ -1,12 +1,93 @@
+~~~~~~~
+
+A device name should represent its physical counterpart - like a multimeter is ``multimeter``.  In general, the term ``recorder`` is used for devices
+that store or generate output (e.g., spike times or synaptic strengths over time). The term  ``generator`` is used for devices that provide input into the simulation.
```
```suggestion
that store the output (e.g., spike times or synaptic strengths over time) of other nodes and make it accessible to the user. The term  ``generator`` is used for devices that provide input into the simulation.
```

akorgor replied:
Do the recorders really generate output? If yes, please ignore my suggestion.

jessica-mitchell commented on doc/extractor_userdocs.py:
```diff
@@ -344,11 +344,17 @@ def mkitem(t):
-        page_title = "Model Directory"
+        page_title = "Model directory"
+        description = """
+                       The models are organized and autogenerated by keywords (e.g., adaptive threshold,
```
Hm good point, I will change the wording to make it clearer.

jessica-mitchell commented on doc/htmldoc/ref_material/glossary.rst:
```diff
@@ -134,6 +88,90 @@ Glossary
+ vp
+   Virtual process.
+
+Commonly used terms in NEST
```
Could you please either create an issue regarding the glossary errors, or make a PR with suggested fixes?

akorgor replied:
Ok

jessica-mitchell commented on doc/htmldoc/models/models-main.rst:
```diff
@@ -1,12 +1,93 @@
+
 NEST provides a ton of models! Textbook standards like integrate-and-fire and Hodgkin-Huxley type models are available
 alongside high quality implementations of models published by the neuroscience community.
+The models are organized and autogenerated by keywords (e.g., :doc:`adaptive threshold <index_adaptive threshold>`, :doc:`conductance-based <index_conductance-based>` etc) in our directory. 
```
fixed

jessica-mitchell commented:
@akorgor thanks for your feedback, I think I addressed all the points you raised, please resolve the remaining conversations and approve if you are satisfied

akorgor commented on doc/htmldoc/models/models-main.rst:
```diff
@@ -1,12 +1,93 @@
+Create and customize models with NESTML
+---------------------------------------
+
+Check out :doc:`NESTML <nestml:index>`, a domain-specific language for neuron and synapse models.
```
What about this comment?

akorgor commented on doc/htmldoc/models/models-main.rst:
```diff
@@ -32,13 +32,6 @@ Check out :doc:`NESTML <nestml:index>`, a domain-specific language for neuron an
 NESTML enables fast prototyping of new models using an easy to understand, yet powerful syntax, and combines
 this with high simulation performance through the automated generation of C++ code, suitable for use in NEST Simulator.
```
Third point got lost.

akorgor commented on doc/htmldoc/models/models-main.rst:
```diff
@@ -32,13 +32,6 @@ Check out :doc:`NESTML <nestml:index>`, a domain-specific language for neuron an
 NESTML enables fast prototyping of new models using an easy to understand, yet powerful syntax, and combines
 this with high simulation performance through the automated generation of C++ code, suitable for use in NEST Simulator.
```
```suggestion
NESTML enables fast prototyping of new models using an easy to understand, yet powerful syntax. This is achieved by a combination of a flexible processing toolchain
written in Python with high simulation performance through the automated generation of C++ code, suitable for use in NEST Simulator.
```

akorgor replied:
The third bulletpoint got lost. Tried to add it back in.
